plugins/ftdidmx/FtdiDmxThread.cpp

```diff
@@ -73,15 +99,164 @@ bool FtdiDmxThread::WriteDMX(const DmxBuffer &buffer) {
+    thread_unmute_callback->Run();
+  } else if (m_branch_callback != nullptr) {
+    thread_branch_callback = m_branch_callback;
+    m_branch_callback = nullptr;
+    thread_branch_callback->Run(nullptr, 0);
+  } else if (m_rdm_callback != nullptr) {
+    thread_rdm_callback = m_rdm_callback;
+    m_rdm_callback = nullptr;
+    ola::rdm::RunRDMCallback(thread_rdm_callback, state);
+  }
+}
+
+void FtdiDmxThread::destroyPendingRequest() {
+  ola::rdm::RDMRequest *tmp_rdm_request = m_pending_request;
+
+  m_pending_request = nullptr;
+  delete tmp_rdm_request;
+}
+
+void FtdiDmxThread::MuteDevice(const ola::rdm::UID &target,
+                               MuteDeviceCallback *mute_complete) {
+  ola::thread::MutexLocker locker(&m_rdm_mutex);
+  if (m_pending_request == nullptr) {
+    OLA_INFO << "Muting device";
+    m_mute_complete = mute_complete;
+    m_pending_request = ola::rdm::NewMuteRequest(m_uid,
+                                                 target,
+                                                 m_transaction_number += 1);
+  } else {
+    // Already pending request
+    OLA_WARN << "Unable to queue Mute request, "
+             << "RDM operation already pending";
+  }
+}
+
+void FtdiDmxThread::UnMuteAll(UnMuteDeviceCallback *unmute_complete) {
+  ola::thread::MutexLocker locker(&m_rdm_mutex);
+  if (m_pending_request == nullptr) {
+    OLA_INFO << "Sending UnMuteAll";
+    m_unmute_complete = unmute_complete;
+    m_pending_request = ola::rdm::NewUnMuteRequest(m_uid,
+                                                   ola::rdm::UID::AllDevices(),
+                                                   m_transaction_number += 1);
+  } else {
+    // Already pending request
+    OLA_WARN << "Unable to queue UnMuteAll request, "
+             << "RDM operation already pending";
+  }
+}
+
+void FtdiDmxThread::Branch(const ola::rdm::UID &lower,
+                           const ola::rdm::UID &upper,
+                           BranchCallback *callback) {
+  ola::thread::MutexLocker locker(&m_rdm_mutex);
+  if (m_pending_request == nullptr) {
+    OLA_INFO << "Sending branch";
+    m_branch_callback = callback;
+    m_pending_request =
+        ola::rdm::NewDiscoveryUniqueBranchRequest(m_uid,
+                                                  lower,
+                                                  upper,
+                                                  m_transaction_number += 1);
+  } else {
+    // Already pending request
+    OLA_WARN << "Unable to queue Branch request, "
+             << "RDM operation already pending";
+  }
+}
 
 /**
  * @brief The method called by the thread
  */
 void *FtdiDmxThread::Run() {
-  TimeStamp ts1, ts2, ts3;
+  OLA_INFO << "Starting FtdiDmxThread";
+  TimeStamp ts1, ts2, ts3, lastDMX;
   Clock clock;
   CheckTimeGranularity();
   DmxBuffer buffer;
+  bool sendRDM = false;
+  TimeInterval elapsed, interval;
+  int readBytes;
+  unsigned int additionalWait = 0;
+  unsigned char readBuffer[258];
```

Comment:
Again where has this magic number come from? I'm guessing it's max RDM packet size + a bit?

Reply:
Yup

Reply:
It's actually - FTDI will always add 1 null byte at the beginning (not 100% sure but I think that is the break being treated as data) of DMX data (except for the DUB reply), the max packet length is 256 bytes and 1 more byte just in case.